### PR TITLE
Performance: Drag items via transform property to avoid layout and re-paints

### PR DESCRIPTION
**Pull request by gwwar**
### Background 

Our draggable component updates a drag chip by modifying `top` and `left` properties. This unfortunately forces the browser to go through layout > paint> and compositing steps which is expensive. As we can see in the profiler on trunk, we're both dropping frames and seeing layout shifts.

Changes in this PR update this to use `transform` instead which can skip layout and paint, [to go straight to composite](https://developers.google.com/web/fundamentals/performance/rendering). We also add a throttle to reduce the number of fired drag events, and early escape if the mouse has not moved.

Our drop zone handlers are also causing major performance issues when dragging quickly. This will be handled in https://github.com/WordPress/gutenberg/pull/33435

Videos below are taken with the Armando theme, using the "alternating rows with images and headings" pattern.

| before | after |
|-----|-----|
| Screen Shot 2021-07-14 at 2 33 23 PM | Screen Shot 2021-07-14 at 2 26 17 PM |

### Testing Instructions

- Verify that there are no regressions with dragging
- When profiling verify that layout shifts are visible in trunk when dragging, but not visible in this branch. (See examples in video).

#### Profiling In Trunk 

- Notice how the layout shifts are coming from the draggable chip.
- The lionshare of handling time is still taken up by dropzone in some cases. This will be addressed in a follow up.

https://user-images.githubusercontent.com/1270189/125692895-71c7d498-928f-43cf-8db4-a4abbf70558c.mp4

#### Profiling in Branch
- No more layout shifts visible from profiler
- Dragging over multiple dropzones is still slow and will be addressed in a follow up. 

https://user-images.githubusercontent.com/1270189/125694389-c3d99e7d-53ed-498d-912b-0e4aa73c0247.mp4

### CODE REVIEW

**Comment by gwwar** [comment]
Playing around with the draggable + dropzone code, it feels like we have a few opportunities.

- We can add a single listener that fires on document dragover to track if our mouse is moving quickly, or not (if the user is hovering). If the user is not hovering we can avoid trying to trigger all these dropzones from rendering/animating.
- Providing a bigger drag hint + drop zone target would be a huge improvement. Maybe https://github.com/WordPress/gutenberg/pull/33132 would help with this.
- Right now **too many dragover events are firing**. I'll continue to look for spots to throttle.
- Since normally users can only drag one item onto a single droppable zone, it may make sense to convert logic to a global listener. The logic is currently a bit hard to follow (and likely more error prone with multiple child listeners).

**Comment by talldan** [comment]
I did some testing and the results seem good so far. 

The blue drop zone indicators are noticeably quite delayed at updating position, so that might be something to improve. I think that's one of those things that needs to feel quite responsive, or the user will perceive it as slow, even if the code is well optimized.

Definitely a good idea to work on this iteratively, so if we can get some noticeable improvements with the Draggable first, we can ship that straight away.

**Comment by gwwar** [comment]
> Definitely a good idea to work on this iteratively, so if we can get some noticeable improvements with the Draggable first, we can ship that straight away.

Sounds good, I can tidy up changes to pull out the draggable changes first (will set this to ready for review when I'm done). General dragging performance is still going to be pretty slow from the existing dropzone handlers, but the difference in removing those layout shifts should be visible from browser profiling.

I'll spin up a follow up draft to continue experimenting with drag intents.

File: packages/components/src/draggable/index.js [hunk 1]
```diff
@@ -129,19 +181,19 @@ export default function Draggable( {
...
-				cloneWrapper.style.transform = 'scale(0.5)';
 				cloneWrapper.style.transformOrigin = 'top left';
 				// Position clone near the cursor.
-				cloneWrapper.style.top = `${ event.clientY - 100 }px`;
-				cloneWrapper.style.left = `${ event.clientX }px`;
+				x = event.clientX;
+				y = event.clientY - 100;
+				// Can also be written as the affine transform: matrix( 0.5, 0, 0, 0.5, x, y ).
+				// Note that order matters, translate should be done before scale.
+				cloneWrapper.style.transform = `translate( ${ x }px, ${ y }px ) scale(0.5)`;
```
Comment (gwwar) [inline packages/components/src/draggable/index.js]: Do we have any examples where we drag an item that's more than 700px? I was wondering if we still needed to handle this case.

**Comment by gwwar** [comment]
This one is ready for review 👍  I moved over other exploratory dropzone performance handling to https://github.com/WordPress/gutenberg/pull/33435

File: packages/components/src/draggable/index.js (hunk 1 above)
Comment (talldan) [reply packages/components/src/draggable/index.js]: I think just about every interaction uses a draggable chip now rather than a clone of an element, so this is just here for consumers of the package.

File: packages/components/src/draggable/index.js [hunk 2]
```diff
@@ -61,6 +82,29 @@ export default function Draggable( {
...
+			dragState.cloneWrapper.style.transform = `translate( ${ nextX }px, ${ nextY }px )`;
+		} else {
+			dragState.cloneWrapper.style.transform = `translate( ${ nextX }px, ${ nextY }px ) scale( ${ dragState.scale } )`;
+		}
+
+		// Update cursor coordinates.
+		dragState.cursorLeft = e.clientX;
+		dragState.cursorTop = e.clientY;
+		dragState.x = nextX;
+		dragState.y = nextY;
```
Comment (talldan) [inline packages/components/src/draggable/index.js]: I wonder if it might be confusing updating the object properties in this way through an object reference. 🤔 

Maybe here it could call an `updateDragState` function that makes the change to a `ref` defined by `useRef` or something. Not sure. I recall the previous version of this code being a bit hard to follow anyhow.

Any other ideas?

File: packages/components/src/draggable/index.js [hunk 3]
```diff
@@ -156,30 +208,37 @@ export default function Draggable( {
...
 
-		ownerDocument.addEventListener( 'dragover', over );
+		// Aim for 60fps (16 ms per frame) for now. We can potentially use requestAnimationFrame (raf) instead,
+		// note that browsers may throttle raf below 60fps in certain conditions.
+		const throttledDragOver = throttle( update, 16 );
+
+		/**
+		 * @param {import('react').DragEvent} e
+		 */
+		ownerDocument.addEventListener( 'dragover', throttledDragOver );
```
Comment (talldan) [inline packages/components/src/draggable/index.js:185]: I think `throttledDragOver` will have to be removed as the event listener on line 270.

**Review by talldan — COMMENTED** [review]
Looks good so far. The removeEventListener issue was the only thing I spotted that seems like it needs to be fixed.

File: packages/components/src/draggable/index.js (hunk 3 above)
Comment (gwwar) [reply packages/components/src/draggable/index.js:185]: Great catch! ✨ Updated in https://github.com/WordPress/gutenberg/pull/33395/commits/4f4294366dda0275ea124e907eb5634d3291af08

File: packages/components/src/draggable/index.js (hunk 1 above)
Comment (gwwar) [reply packages/components/src/draggable/index.js]: I'm quite tempted to rip out this case. Here's what it looks like in **trunk** with a draggable with a 701x701px square. Scaling to 50% of the original draggable isn't going to work with larger items, so folks will likely need to come up with a similar solution to what we went with.

The positioning also looks off, cursor at top left vs center (which I can fix, but would modify current behavior).

https://user-images.githubusercontent.com/1270189/125821948-92485ce6-4f6a-4907-96d8-90ce5355680e.mp4

Comment (gwwar) [reply packages/components/src/draggable/index.js]: I tried removing this case in https://github.com/WordPress/gutenberg/pull/33395/commits/9667539e3a8f1dfb31a659c04a274444bd8fe1cb

File: packages/components/src/draggable/index.js (hunk 2 above)
Comment (gwwar) [reply packages/components/src/draggable/index.js]: I tried removing the dragState object in https://github.com/WordPress/gutenberg/pull/33395/commits/9667539e3a8f1dfb31a659c04a274444bd8fe1cb removing the extra function.

Closures are pretty common in animation/drawing handlers (since we need to keep track of what happened in the last step vs the current step), but we could try to tidy up function signatures in a larger refactor. I'd maybe move that to a seperate PR though.

File: packages/components/src/draggable/index.js (hunk 1 above)
Comment (talldan) [reply packages/components/src/draggable/index.js]: Hmm, yeah. It feels like it'd be better if it had a max size instead of reducing by 50%, but maybe that's to stop the drag image from being blurry.

Removing it seems ok, we can monitor to see if there's any negative feedback.

I imagine this was for large blocks, so the use case is more than likely no longer present.

**Review by talldan — APPROVED** [review]
Thanks for working on this, lets get this in.
